golang/cosmos/x/swingset/genesis.go

```diff
@@ -169,32 +190,73 @@ type swingStoreGenesisEventHandler struct {
 	snapshotHeight uint64
 	swingStore     sdk.KVStore
 	hasher         hash.Hash
+	exportMode     string
 }
 
 func (eventHandler swingStoreGenesisEventHandler) OnExportStarted(height uint64, retrieveSwingStoreExport func() error) error {
 	return retrieveSwingStoreExport()
 }
 
 func (eventHandler swingStoreGenesisEventHandler) OnExportRetrieved(provider keeper.SwingStoreExportProvider) error {
-	if eventHandler.snapshotHeight != provider.BlockHeight {
+	if !((eventHandler.exportMode == keeper.SwingStoreArtifactModeDebug && eventHandler.snapshotHeight == 0) ||
+		eventHandler.snapshotHeight == provider.BlockHeight) {
 		return fmt.Errorf("snapshot block height (%d) doesn't match requested height (%d)", provider.BlockHeight, eventHandler.snapshotHeight)
 	}
 
+	artifactsEnded := false
+
+	var getExportDataReader = func() (agoric.KVEntryReader, error) {
+		exportDataIterator := eventHandler.swingStore.Iterator(nil, nil)
+		kvReader := agoric.NewKVIteratorReader(exportDataIterator)
+		eventHandler.hasher.Reset()
+		encoder := json.NewEncoder(eventHandler.hasher)
+		encoder.SetEscapeHTML(false)
+
+		return agoric.NewKVHookingReader(kvReader, func(entry agoric.KVEntry) error {
+			return encoder.Encode(entry)
+		}, func() error {
+			return nil
+		}), nil
+	}
+
 	artifactsProvider := keeper.SwingStoreExportProvider{
-		GetExportDataReader: func() (agoric.KVEntryReader, error) {
-			exportDataIterator := eventHandler.swingStore.Iterator(nil, nil)
-			kvReader := agoric.NewKVIteratorReader(exportDataIterator)
-			eventHandler.hasher.Reset()
-			encoder := json.NewEncoder(eventHandler.hasher)
-			encoder.SetEscapeHTML(false)
+		GetExportDataReader: getExportDataReader,
+		ReadNextArtifact: func() (types.SwingStoreArtifact, error) {
+			var (
+				artifact          types.SwingStoreArtifact
+				err               error
+				encodedExportData bytes.Buffer
+			)
 
-			return agoric.NewKVHookingReader(kvReader, func(entry agoric.KVEntry) error {
-				return encoder.Encode(entry)
-			}, func() error {
-				return nil
-			}), nil
+			if !artifactsEnded {
+				artifact, err = provider.ReadNextArtifact()
+			} else {
+				return types.SwingStoreArtifact{}, io.EOF
+			}
+
+			if err == io.EOF {
+				artifactsEnded = true
+				if eventHandler.exportMode == keeper.SwingStoreArtifactModeDebug {
+					err = nil
+					exportDataReader, err := getExportDataReader()
```

Comment:
I thought `EncodeKVEntryReaderToJsonl` function will close the reader passed to it at the end but it seems like it doesn't

Reply:
Right, `EncodeKVEntryReaderToJsonl` doesn't close the reader. I think that was because I wanted the reader to be closed even if a read error occurred, and leave the caller in charge of handling potentially double errors. I'm not sure if that's the idiomatic golang thing to do, and would be open to reconsider if it's not.